native/c/zdsm.c

```diff
@@ -95,6 +95,93 @@ typedef int (*IGWASMS)(
     int *ds_type // 1 for PDSE, 2 for HFS
     ) ATTRIBUTE(amode31);
 
+#pragma prolog(ZDSDSORG, " ZWEPROLG NEWDSA=(YES,128) ")
+#pragma epilog(ZDSDSORG, " ZWEEPILG ")
+
+// Obtain the data set organization for a data set, given its name and volser
+// Full PDF for DFSMSdfp advanced services: https://www.ibm.com/docs/en/SSLTBW_2.5.0/pdf/idas300_v2r5.pdf
+// Doc page: https://www.ibm.com/docs/en/zos/3.1.0?topic=macros-reading-dscbs-from-vtoc-using-obtain
+int ZDSDSORG(ZDS *zds, const char *dsn, const char *volser, char *dsorg_buf,
+             int dsorg_buf_len)
+{
+  // workarea: each DSCB is 140 bytes, we need enough space for format-1 DSCB, format-8 DSCB and max possible format-3 DSCBs (10)
+  // adding 140 bytes for the workarea itself
+  char workarea[sizeof(IndexableDSCBFormat1) * MAX_DSCBS];
+  ObtainParams params = {0};
+
+  // We're using OBTAIN through CAMLST SEARCH to get the DSCBs, see here for more info:
+  // https://www.ibm.com/docs/en/zos/3.1.0?topic=obtain-reading-dscb-by-data-set-name
+
+  // OBTAIN by data set name
+  params.reserved = 0xC1;
+  params.number_dscbs = MAX_DSCBS;
+  params.option_flags = OPTION_EADSCB;
+  // Allow lookup of format-1 or format-8 DSCB
+  char dsname[44] = {0};
+  memset(dsname, ' ', sizeof(dsname));
+  memcpy(dsname, dsn, strlen(dsn));
+  params.listname_addrx.dsname_ptr = dsname;
+  char volume[6] = {0};
+  memset(volume, ' ', sizeof(volume));
+  memcpy(volume, volser, strlen(volser));
+  params.listname_addrx.volume_ptr = volume;
+  params.listname_addrx.workarea_ptr = workarea;
+
+  int rc = obtain_camlst(params);
+  if (0 != rc)
+  {
+    strcpy(zds->diag.service_name, "OBTAIN");
+    zds->diag.e_msg_len =
+        sprintf(zds->diag.e_msg, "OBTAIN SVC failed for %s on %s with rc=%d, workarea_ptr=%p",
+                dsn, volser, rc, workarea);
+    zds->diag.service_rc = rc;
+    zds->diag.detail_rc = ZDS_RTNCD_SERVICE_FAILURE;
+    return RTNCD_FAILURE;
+  }
+
+  IndexableDSCBFormat1 indexable_dscb = {0};
+  for (int i = 0; i < MAX_DSCBS - 1; i++)
+  {
+    memcpy(&indexable_dscb, workarea + (i * (sizeof(IndexableDSCBFormat1) - 1)), sizeof(indexable_dscb));
+    // The returned DSCB does not include the key, but we can infer the returned variables by re-aligning the struct
+    DSCBFormat1 *dscb = (DSCBFormat1 *)&indexable_dscb;
+
+    // '1' or '8' in EBCDIC
```

Comment:
what is special about a 1 or 8?

Reply:
I believe we are looking for either a Format-1 and Format-8 DSCB, which are formatted similarly

Reply:
You are not dense 😋  Addressed in https://github.com/zowe/zowe-native-proto/pull/510/commits/3905fec0efe51a672239082f777ad87e4ca20944